### PR TITLE
[updates] introduce overrideConfiguration

### PR DESCRIPTION
**Pull request by Kudo**
# Why

#26078 for sdk 50 and main

# How

- [android] try to port #26078 to the new UpdatesController.
- [android] deprecate the original `UpdatesController.initialize(context, configuration)`. it will have a compile error anyway.
- [ios] add AppController.overrideConfiguration()
- [doc] update doc for the new `overrideConfiguration`

# Test Plan

- on android follows #26078 steps on a sdk 50 project
- on ios, it's pretty the same but to
  - disable c++/swift optimization for release build, so that we can set breakpoints
  - patch AppDelegate.mm to add override headers
```diff
--- a/ios/sdk50/AppDelegate.mm
+++ b/ios/sdk50/AppDelegate.mm
@@ -3,6 +3,10 @@
 #import <React/RCTBundleURLProvider.h>
 #import <React/RCTLinkingManager.h>

+#import "ExpoModulesCore-Swift.h"
+#import "EXUpdatesInterface-Swift.h"
+#import "EXUpdates-Swift.h"
+
 @implementation AppDelegate

 - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
@@ -13,6 +17,13 @@
   // They will be passed down to the ViewController used by React Native.
   self.initialProps = @{};

+  NSDictionary *overrideConfig = @{
+    @"EXUpdatesRequestHeaders": @{
+      @"Foo": @"Bar"
+    }
+  };
+  [EXUpdatesAppController overrideConfigurationWithConfiguration:overrideConfig];
+
   return [super application:application didFinishLaunchingWithOptions:launchOptions];
 }
```

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: docs/pages/versions/unversioned/sdk/updates.mdx [hunk 1]
```diff
@@ -105,9 +105,15 @@ Every custom updates server must implement the [Expo Updates protocol](/technica
...
+On Android, these options are set as `meta-data` tags adjacent to the tags added during installation in the **AndroidManifest.xml** file. You can also define these options at runtime by passing a `Map` as the second parameter of `UpdatesController.overrideConfiguration()`, and the provided values will override any value specified in **AndroidManifest.xml**.
 
-On iOS, these properties are set as keys in **Expo.plist** file. You can also set them at runtime by calling `[EXUpdatesAppController.sharedInstance setConfiguration:]` at any point before calling `start` or `startAndShowLaunchScreen`, and the values in this dictionary will override any values specified in **Expo.plist**.
+On iOS, these properties are set as keys in **Expo.plist** file. You can also set them at runtime by calling `[EXUpdatesAppController overrideConfigurationWithConfiguration:]` at any point before `expo-updates` initialized, and the values in this dictionary will override any values specified in **Expo.plist**. If you tried to use the `EXUpdatesAppController` from the **AppDelegate.mm**, you may need to add the following imports:
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/sdk/updates.mdx:110]:
```suggestion
On iOS, these properties are set as keys in the **Expo.plist** file. You can also set them at runtime by calling `[EXUpdatesAppController overrideConfigurationWithConfiguration:]` at any point before `expo-updates` is initialized, and the values in this dictionary will override any values specified in **Expo.plist**. If you tried to use the `EXUpdatesAppController` from the **AppDelegate.mm**, you will need to add the following imports:
```

Small grammatical corrections.

File: docs/pages/versions/v50.0.0/sdk/updates.mdx [hunk 2]
```diff
@@ -105,9 +105,16 @@ Every custom updates server must implement the [Expo Updates protocol](/technica
...
 
-On Android, these options are set as `meta-data` tags adjacent to the tags added during installation in the **AndroidManifest.xml** file. You can also define these options at runtime by passing a `Map` as the second parameter of `UpdatesController.initialize()`, and the provided values will override any value specified in **AndroidManifest.xml**.
+On Android, these options are set as `meta-data` tags adjacent to the tags added during installation in the **AndroidManifest.xml** file. You can also define these options at runtime by passing a `Map` as the second parameter of `UpdatesController.overrideConfiguration()`, and the provided values will override any value specified in **AndroidManifest.xml**.
+
+On iOS, these properties are set as keys in **Expo.plist** file. You can also set them at runtime by calling `[EXUpdatesAppController overrideConfigurationWithConfiguration:]` at any point before `expo-updates` initialized, and the values in this dictionary will override any values specified in **Expo.plist**. If you tried to use the `EXUpdatesAppController` from the **AppDelegate.mm**, you may need to add the following imports:
```
Comment (amandeepmittal) [inline docs/pages/versions/v50.0.0/sdk/updates.mdx:110]:
```suggestion
On iOS, these properties are set as keys in the **Expo.plist** file. You can also set them at runtime by calling `[EXUpdatesAppController overrideConfigurationWithConfiguration:]` at any point before `expo-updates` is initialized, and the values in this dictionary will override any values specified in **Expo.plist**. If you tried to use the `EXUpdatesAppController` from the **AppDelegate.mm**, you will need to add the following imports:
```

**Review by amandeepmittal — APPROVED** [review]

**Review by douglowder — APPROVED** [review]

**Comment by douglowder** [comment]
@Kudo I'm seeing the exact same failures in the iOS E2E tests (debug build) after two test runs. It seems there is some issue where restarting the app does not successfully download the new update:

Screenshot 2023-12-24 at 10 34 09 AM

**Comment by Kudo** [comment]
> @Kudo I'm seeing the exact same failures in the iOS E2E tests (debug build) after two test runs. It seems there is some issue where restarting the app does not successfully download the new update:

that's strange. i didn't see any relevant code for the failure. from the latest changes, that tests are just passed.

**Review by wschurman — APPROVED** [review]

**Pull request merged**